File: Parallel-Csharp/src/Parallel/Linq101/QueryExecution.cs
using System;
using System.Linq;

namespace Parallel.Linq101;

public class QueryExecution {
    public int DeferredExecution() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int i = 0;
        var q = numbers.Select(n => ++i);
        foreach (var v in q)
            Console.WriteLine($"v = {v}, i = {i}");
        return 0;
    }

    public int EagerExecution() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        int i = 0;
        var q = numbers.Select(n => ++i).ToArray();
        foreach (var v in q)
            Console.WriteLine($"v = {v}, i = {i}");
        return 0;
    }

    public int ReuseQueryDefinition() {
        int[] numbers = { 5, 4, 1, 3, 9, 8, 6, 7, 2, 0 };
        var lowNumbers = numbers.Where(n => n <= 3);
        var lowNumbers2 = numbers.Where(n => n <= 3);
        Console.WriteLine("First run numbers <= 3:");
        foreach (int n in lowNumbers)
            Console.WriteLine(n);
        for (int i = 0; i < 10; i++)
            numbers[i] = -numbers[i];
        Console.WriteLine("Second run numbers <= 3:");
        foreach (int n in lowNumbers2)
            Console.WriteLine(n);
        return 0;
    }
}

File: Parallel-Java/src/parallel/linq101/QueryExecution.java
package parallel.linq101;

import java.util.Arrays;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.stream.IntStream;

public class QueryExecution {
    public int deferredExecution() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        AtomicInteger i = new AtomicInteger();
        var q = Arrays.stream(numbers).map(n -> i.incrementAndGet());
        q.forEach(v -> System.out.println("v = " + v + ", i = " + i));
        return 0;
    }

    public int eagerExecution() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        AtomicInteger i = new AtomicInteger();
        var q = Arrays.stream(numbers).map(n -> i.incrementAndGet()).toArray();
        for (var v : q)
            System.out.println("v = " + v + ", i = " + i);
        return 0;
    }

    public int reuseQueryDefinition() {
        int[] numbers = {5, 4, 1, 3, 9, 8, 6, 7, 2, 0};
        var lowNumbers = Arrays.stream(numbers).filter(n -> n <= 3);
        var lowNumbers2 = Arrays.stream(numbers).filter(n -> n <= 3);
        System.out.println("First run numbers <= 3:");
        lowNumbers.forEach(n -> System.out.println(n));
        for (int i = 0; i < 10; i++)
            numbers[i] = -numbers[i];
        System.out.println("Second run numbers <= 3:");
        lowNumbers2.forEach(n -> System.out.println(n));
        return 0;
    }
}
